Make Kazoo work without gevent (for example in sidecars)

RoganMurley opened the pull request:
## 💸 TL;DR
KazooClient needs to be initialized differently depending on whether we are using gevent or not. We don't use gevent in sidecars so need to make sure KazooClient works without gevent.

## 📜 Details
In #659 we updated Kazoo to use `SequentialGeventHandler` for compatibility with gevent. Unfortunately this handler silently fails if we are not using gevent.

Sidecars do _not_ use gevent as they don't use entrypoint scripts that patch (such as `baseplate-serve`), instead being ran directly (`python -m baseplate.sidecars.live_data_watcher some_config.ini`). I'm not sure if this is intentional or an oversight.

This PR detects if gevent patching has occurred and initializes KazooClient accordingly.

## 🧪 Testing Steps / Validation
* This change has been tested on an internal service that was experiencing the above issue when updating to the latest baseplate version.

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [x] CI tests (if present) are passing
- [x] Adheres to code style for repo
- [x] Contributor License Agreement (CLA) completed if not a Reddit employee

Commit:
Initialize KazooClient differently depending on gevent

Commit:
Fix lint

Commit:
Fix test

Commit:
Add kazoo with and without gevent tests

SiddharthManoj commented:
I verified that this resolves the issue of the ZK source file not triggering [live updates](https://github.com/reddit/baseplate.py/blob/develop/baseplate/sidecars/live_data_watcher.py#L118) (`on_change`). I verified locally by doing the following:

Creds to @TylerLubeck for tips on how to spin up a local testing env as follows:

1. Make a venv

2. Inside the venv 
`pip3 install -e .[zookeeper,s3fetcher]`

3. make a config file (example below) live-data.ini in the directory:

```
[app:main]
factory = helloworld:make_wsgi_app

metrics.tagging = true
metrics.log_if_unconfigured = true

db.url = sqlite:///

[live-data]
zookeeper.hosts = localhost:2181

nodes.a.source = /data-three
nodes.a.dest = /tmp/data-three

[server:main]
```

4. run `docker run --name zk -p 2181:2181 zookeeper`

5. create a znode `docker exec zk zkCli.sh create /data`

6. update the znode (do this repeatedly) `docker exec zk zkCli.sh set /data new-data-or-whatever-you-want`

7. Change permissions for secrets:  `sudo mkdir -p /var/local;  sudo chown -R ${USER}:staff /var/local; echo "{}" > /var/local/secrets.json`

8. Run the live-data-watcher `python -m baseplate.sidecars.live_data_watcher live-data.ini`

KTAtkinson commented on baseplate/lib/live_data/zookeeper.py:
```diff
@@ -56,12 +59,21 @@ def zookeeper_client_from_config(
 
+    # Kazoo requires different parameters depending on whether
+    # we are using gevent or not.
+    if is_gevent_patched():
```
This should probably be called `gevent_is_patched`

KTAtkinson commented on baseplate/server/monkey.py:
```diff
@@ -10,4 +10,9 @@ def patch_stdlib_queues() -> None:
+
+
+def is_gevent_patched() -> bool:
+    """Has any gevent patching has occurred?"""
```
Please also describe what the return value is for this function.

RoganMurley replied:
I've added a cast to make sure that it's actually a `bool`.

Do I need to add a return type to the docstring even though I have a `bool` type annotation on the function?

KTAtkinson reviewed (changes requested):
The PR description says that you intend to use this for sidecars so that they don't try to uses gevent, but I don't see any code related to updating how the sidecars work. Is this intended?

Commit:
Address review feedback

RoganMurley commented:
> The PR description says that you intend to use this for sidecars so that they don't try to uses gevent, but I don't see any code related to updating how the sidecars work. Is this intended?

To clarify, sidecars currently do **not** use gevent because they aren't run with an entrypoint script that does gevent patching ([see docs](https://baseplate.readthedocs.io/en/stable/api/baseplate/lib/live_data.html#watcher-daemon)). This surprised me and I think that there's a discussion to be had about whether to use gevent with them, but that would be a bigger change.

This PR detects if gevent patching has occurred and chooses the correct Kazoo handler accordingly. This means that the KazooClient can be safely used in gevent environments (`baseplate-serve`, `baseplate-script`) and non-gevent environments (`sidecars`).

nsheaps commented on baseplate/lib/live_data/zookeeper.py:
```diff
@@ -56,12 +59,21 @@ def zookeeper_client_from_config(
+        sleep_func = gevent.sleep
+    else:
+        handler = None
+        sleep_func = time.sleep
```
If gevent does it's [patching this becomes gevent.sleep anyway](https://www.gevent.org/api/gevent.monkey.html#gevent.monkey.patch_time). Do we need to explicitly use one or the other here or can we just use time.sleep always?

RoganMurley replied:
You're right, this can be removed!

I added it because of a failing test, but in retrospect that was because the sleep function hadn't been patched in that test.

nsheaps commented on baseplate/lib/live_data/zookeeper.py:
```diff
@@ -56,12 +59,21 @@ def zookeeper_client_from_config(
+    # Kazoo requires different parameters depending on whether
+    # we are using gevent or not.
+    if gevent_is_patched():
+        handler = SequentialGeventHandler()
```
This feels like something we either want or don't want all the time regardless of if gevent has patched items

RoganMurley replied:
Kazoo intentionally eschews monkey patching, see the documentation below. It expects you to explicitly choose the handler yourself if you use gevent.

https://kazoo.readthedocs.io/en/latest/async_usage.html#connection-handling
```
Kazoo doesn’t rely on gevents/eventlet monkey patching and requires that you pass in the appropriate handler, the default handler is SequentialThreadingHandler
```

Kazoo docs have pretty strong guidance about not using `SequentialThreadingHandler` with gevent.

https://kazoo.readthedocs.io/en/2.6.1/api/handlers/threading.html?highlight=sequentialthreadinghandler#kazoo-sasl-handlers-threading
```
Do not use SequentialThreadingHandler with applications using asynchronous event loops (like gevent). Use the SequentialGeventHandler instead.
```

My understanding from reading the [code](https://kazoo.readthedocs.io/en/2.6.1/_modules/kazoo_sasl/handlers/threading.html#SequentialThreadingHandler) and [GitHub issues](https://github.com/python-zk/kazoo/issues/3#issuecomment-7102481) is that the Zookeeper C client library is not patched by gevent and so [spawns a separate thread](https://gevent.narkive.com/Vn0vZP9R/is-it-possible-to-execute-code-in-main-thread-when-called-from-another-thread). However, this is not well documented and we haven't been _knowingly_ hurt by it.

`SequentialThreadingHandler` _did_ seem to work okay before we patched gevent queues in #659, where we found that the handler used queue methods not available on gevent queues. This brought `SequentialGeventHandler` to our attention, which was designed to work with gevent queues from the start.

I think using an implementation built for explicitly for gevent rather than a monkey patched one is a good thing. Not only will we avoid issues with the C lib that we don't even know about, gevent code can have different semantics and performance characteristics.

nsheaps commented on baseplate/server/monkey.py:
```diff
@@ -10,4 +10,9 @@ def patch_stdlib_queues() -> None:
+
+def gevent_is_patched() -> bool:
+    """Has any gevent patching has occurred?"""
+    return bool(monkey.saved)
```
**any** doesn't necessarily mean that it has patched what we want. What exactly does "use gevent and this works or don't and it silently fails" relate back to the usage of?

RoganMurley replied:
I'm not actually sure... it's likely the case that an unpatched operation is yielded to that never yields back to gevent because it's not patched to do so. This makes it difficult to point to a single unpatched lib.

I'm not too worried about supporting beyond `patch_all` given that we never patch granularly outside of tests.

nsheaps reviewed (changes requested):
If patching occurs then in theory you shouldn't need to check, you can just use the native handlers.

That said, I feel like we should be committing to gevent because that's what we want, or fall back on native handling because we want patching to replace it when it happens.

I think I need a bit more info on what the silent failure mode you described in the PR description is before I have enough context to thoughtfully approve this.

RoganMurley commented:
> If patching occurs then in theory you shouldn't need to check, you can just use the native handlers.
> 
> That said, I feel like we should be committing to gevent because that's what we want, or fall back on native handling because we want patching to replace it when it happens.
> 
> I think I need a bit more info on what the silent failure mode you described in the PR description is before I have enough context to thoughtfully approve this.

Thanks for your feedback, I have addressed some of your comments inline.

I think that committing to gevent is probably the right thing to do in the long term and would be interested in doing that migration. However, I've been bitten enough by weird gevent patching to not want to do it as part of this bugfix. I think we should do a controlled prerelease of any such change, canarying a gevent patching entrypoint before general release.

Commit:
Rely on patching time.sleep

RoganMurley commented on tests/unit/server/monkey_tests.py:
```diff
@@ -2,15 +2,24 @@
 
 
 class MonkeyPatchTests(unittest.TestCase):
+    def tearDown(self):
```
I moved this cleanup to the teardown because otherwise if this test fails it can interfere with other tests.

RoganMurley commented on baseplate/lib/live_data/zookeeper.py:
```diff
@@ -83,6 +90,6 @@ def zookeeper_client_from_config(
             max_jitter=1,  # maximum amount to jitter sleeptimes
             max_delay=60,  # never wait longer than this
-            sleep_func=gevent.sleep,
+            sleep_func=time.sleep,
```
We could just rely on the default value, but passing in time.sleep makes mocking much easier.

nsheaps replied:
I would agree, if this isn't different from the defaults it doesn't make much sense to pass it.

KTAtkinson replied:
:no_bell: Can you add a comment to this effect here?

RoganMurley replied:
@nsheaps  I can do that, but the patching gets much nastier because default values are not easily patched 🤢 I just pushed a commit with that change so let me know what you think

nsheaps replied:
if passing time.sleep makes this all easier, feel free to pass it. If we want to be sticklers about it, then max_delayu is also defaulting to 60, and backoff to 2 ( [source](https://github.com/python-zk/kazoo/blob/master/kazoo/retry.py#L45) )

RoganMurley commented:
@nsheaps I've pushed some changes to rely on the patching `time.sleep`. I had to fiddle with the tests a bit, the comments should explain why 😓

nsheaps commented on tests/integration/live_data/zookeeper_tests.py:
```diff
@@ -48,3 +53,37 @@ def test_create_client_with_credentials(self):
+class ZooKeeperGeventTests(unittest.TestCase):
+    def setUp(self):
+        # We patch `socket` just to make sure that the gevent handler is chosen.
+        # Nothing is special about `socket` in particular.
```
If we don't care about socket in particular, I would suggest patching the same way baseplate does, aka patch_all + baseplate.monkey.patch_queue

RoganMurley replied:
We can't do this because some of what `patch_all` patches we can't "unpatch" (as far as I know). For example, we have no way to unpatch `time.sleep` as it is a builtin and `importlib.reload` doesn't unload it.

RoganMurley replied:
(I'll add a comment to this effect)

RoganMurley replied:
It might make sense to review how we handle gevent tests in a future PR as the current state of things is very fiddly.

nsheaps replied:
https://docs.python.org/3/library/importlib.html#importlib.reload

You might just have to import `time` before patching

nsheaps commented on tests/integration/live_data/zookeeper_tests.py:
```diff
@@ -48,3 +53,37 @@ def test_create_client_with_credentials(self):
         secrets.get_simple.assert_called_with("secret/zk-user")
         self.assertEqual(list(client.auth_data), [("digest", "myzkuser:hunter2")])
+
+    def test_create_client_uses_correct_handler(self):
```
🔕  by the existence of the other test class, I think it makes sense to move it to the other one and drop the word "correct" as both are correct depending on the use case. `test_client_uses_gevent_handler_when_queue_patched` helps identify why it should use gevent handler and differentiates it from the other test.

Since the other class has only 1 test in it, you should move the patching to each test so the setUp doesn't impact this test.

Similarly the other class should be named something more like `ZooKeeperHandlerTestsWithPatching`

This is probably one of the few cases that the more "enterprise-y" the name the easier it is for the next person to reason about the test

RoganMurley replied:
Fair, I suppose the `tearDown` won't affect the other cases so there's no need to be separate.

I've gone with the name `ZooKeeperHandlerWithPatchingTests` to follow the `.*Tests` pattern.

nsheaps reviewed (changes requested):
Few more changes

KTAtkinson commented on baseplate/server/monkey.py:
```diff
@@ -10,4 +10,9 @@ def patch_stdlib_queues() -> None:
+
+
+def gevent_is_patched() -> bool:
+    """Has any gevent patching has occurred?"""
```
```suggestion
    """Returns true if gevent is patched, otherwise returns false."""
```

Commit:
Address review feedback

Commit:
Mock default time.sleep kwarg

RoganMurley commented on tests/integration/live_data/zookeeper_tests.py:
```diff
@@ -48,3 +60,30 @@ def test_create_client_with_credentials(self):
+
+    @mock.patch(
+        "kazoo.retry.KazooRetry.__init__.__defaults__",
+        (1, 0.1, 2, 0.4, 60.0, True, gevent.sleep, None, None),
```
@nsheaps Here is the yucky patch. There are some tricks I could do with a context manager to only modify the one default arg, but I'm unsure if it's worth it.

nsheaps replied:
💡  If you're looking to unit test this, why not mock `baseplate.server.monkey.gevent_is_patched` and test which instance is in the handler without looking to actually monkey patch things?

RoganMurley replied:
Unfortunately the Gevent handler throws a configuration error if the sleep function is unpatched, so we'd still need to patch `time.sleep` in this case.

```
kazoo.exceptions.ConfigurationError: Retry handler and event handler  must use the same sleep func
```

We could patch `gevent_is_patched` if we wanted to avoid doing any gevent patching in test though, that might make things cleaner.

Commit:
Improve patching

RoganMurley commented:
@nsheaps @KTAtkinson I tidied up the testing a bit:
* Mocked `gevent_is_patched` rather than do a true gevent patched test.
* I worked out how to patch default values in a less brittle way.

Commit:
Fix tests

nsheaps commented on tests/integration/live_data/zookeeper_tests.py:
```diff
@@ -1,9 +1,15 @@
+    from kazoo.handlers.gevent import SequentialGeventHandler
+    from kazoo.handlers.threading import SequentialThreadingHandler
+    from kazoo.retry import KazooRetry
 except ImportError:
```
Out of the scope of this PR since you didn't change it, but this seems like something that shouldn't be skipped if something is not installed, especially for tests.

nsheaps commented on tests/integration/live_data/zookeeper_tests.py:
```diff
@@ -48,3 +54,23 @@ def test_create_client_with_credentials(self):
+        )
+        assert isinstance(client.handler, SequentialThreadingHandler)
+
+    @mock.patch("baseplate.lib.live_data.zookeeper.gevent_is_patched", return_value=True)
```
🔕 You should patch this on the other test too so that your results will always be consistent when someone else messes up their tests and accidentally patches without cleaning it up.

nsheaps reviewed (approved):
Nice job. Thanks for addressing all of our feedback!
